src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,286 @@
+  }
+  m_capacity -= 1;
+  if (!tryInsert(m_buckets, reinterpret_cast<uintptr_t>(pointer), size)) {
+    allocerr("failed to insert allocation into allocation table");
+  }
+}
+
+void* rho::AllocationTable::lookup_pointer(uintptr_t candidate) {
+  Allocation* bucket = search(candidate);
+  if (bucket) {
+    if (bucket->isSuperblock()) {
+      AllocatorSuperblock* superblock = bucket->asSuperblock();
+      return superblock->lookupBlock(candidate);
+    } else {
+      return bucket->asPointer();
+    }
+  }
+  return nullptr;
+}
+
+void rho::AllocationTable::erase(uintptr_t pointer, size_t size) {
+  pointer &= ~uintptr_t{3};  // Mask out the flag bits for equality comparison.
+  size_t key_end = endKey(pointer, size);
+  bool first = true;
+  for (size_t key = startKey(pointer); key < key_end; ++key) {
+    size_t hash = pointerHash(key << LOW_BITS);
+    for (int i = 0; i < m_num_buckets; ++i) {
+      Allocation& bucket = m_buckets[hash];
+      if (bucket.dataPointer() == pointer) {
+        m_capacity += 1;
+        bucket.m_data = Allocation::s_deleted_key;
+        break;
+      }
+      if (bucket.isEmpty()) {
+        // No more partial entries inserted, we are done.
+        return;
+      }
+      m_num_erase_collisions += 1;
+      hash = probeFunc(hash, i);
+    }
+  }
+}
+
+rho::AllocationTable::Allocation* rho::AllocationTable::search(uintptr_t candidate) {
+  size_t hash = pointerHash(candidate);
+  for (int i = 0; i < m_num_buckets; ++i) {
+    Allocation* bucket = &m_buckets[hash];
+    if (!bucket->isEmpty() && !bucket->isDeleted()) {
+      if (bucket->isSuperblock()) {
+        AllocatorSuperblock* superblock = bucket->asSuperblock();
+        if (superblock->isInternalPointer(candidate)) {
+          // Pointer is inside this superblock. Test if it is after the header.
+          if (superblock->isBlockPointer(candidate)) {
+            return bucket;
+          } else {
+            return nullptr;
+          }
+        }
+      } else if (bucket->isInternalPointer(candidate)) {
+        return bucket;
+      }
+    }
+    if (bucket->isEmpty()) {
+      return nullptr;
+    }
+    m_num_search_collisions += 1;
+    hash = probeFunc(hash, i);
+  }
+  return nullptr;
+}
+
+bool rho::AllocationTable::resize(unsigned new_alloctable_bits) {
+  unsigned new_table_size = 1 << new_alloctable_bits;
+  unsigned old_mask = m_hash_mask;
+  unsigned new_capacity = new_table_size / 2;  // Sets max load factor to 50%.
+  m_hash_mask = new_table_size - 1;
+  Allocation* new_table = new Allocation[new_table_size];
+  // Build new table.
+  for (int i = 0; i < m_num_buckets; ++i) {
+    Allocation& bucket = m_buckets[i];
+    if (!bucket.isEmpty() && !bucket.isDeleted() && bucket.isFirst()) {
+      // This is the initial hash bucket for the corresponding allocation.
+      // Rebuild the hash entries in the new table based only on this entry.
+      new_capacity -= 1;
+      if (!tryInsert(new_table, bucket.m_data & ~uintptr_t{1},
```

Comment:
Since tryInsert never needs to fail, this entire function can be simplified a bunch.

If that wasn't the case, you could simplify as:
    AllocationTable* new_table = new AllocationTable(new_allocatable_bits);
    ... copy over the entries or delete new_table and exit.
    std::swap(*this, *new_table);